astroquery/mast/cloud.py

```diff
@@ -156,8 +139,32 @@ def get_cloud_uri_list(self, data_products, include_bucket=True, full_url=False)
             List of URIs generated from the data products, list way contain entries that are None
             if data_products includes products not found in the cloud.
         """
+        s3_client = self.boto3.client('s3', config=self.config)
 
-        return [self.get_cloud_uri(product, include_bucket, full_url) for product in data_products]
+        paths = utils.mast_relative_path(data_products["dataURI"])
+        if isinstance(paths, str):  # Handle the case where only one product was requested
+            paths = [paths]
+
+        uri_list = []
+        for path in paths:
+            if path is None:
+                uri_list.append(None)
+            else:
+                try:
+                    # Use `head_object` to verify that the product is available on S3 (not all products are)
+                    s3_client.head_object(Bucket=self.pubdata_bucket, Key=path)
+                    if include_bucket:
+                        path = "s3://{}/{}".format(self.pubdata_bucket, path)
```

Comment:
I don't necessarily like the overwriting of variables here as when this has to be debugged it's a bit more difficult to distinguish between the `path` from the iteration vs the path here. Maybe use `s3_path`?